lib/flowViewport/API/perViewportSceneIndicesData/fvpViewportInformationAndSceneIndicesPerViewportDataManager.cpp

```diff
@@ -0,0 +1,180 @@
+{
+    //Add it in our array if it is not already inside
+    {
+        std::lock_guard<std::mutex> lock(_viewportInformationAndSceneIndicesPerViewportDataSet_mutex);
+
+        auto findResult = std::find_if(_viewportInformationAndSceneIndicesPerViewportDataSet.begin(), _viewportInformationAndSceneIndicesPerViewportDataSet.end(),
+                    [&viewportInfo](const ViewportInformationAndSceneIndicesPerViewportData* other) { 
+                        return &(other->GetViewportInformation()) == &viewportInfo;
+                    }
+        );
+        if (findResult == _viewportInformationAndSceneIndicesPerViewportDataSet.end()){
+            const ViewportInformationAndSceneIndicesPerViewportData::CreationParameters creationParams(viewportInfo, renderIndexProxy);
+            _viewportInformationAndSceneIndicesPerViewportDataSet.insert(new ViewportInformationAndSceneIndicesPerViewportData(creationParams));
+        }else{
+            return; //It is already inside our array
+        }
+    }
+
+    //Let the registered clients know a new viewport has been added
+    InformationInterfaceImp::Get().SceneIndexAdded(viewportInfo);
+}
+
+void ViewportInformationAndSceneIndicesPerViewportDataManager::RemoveViewportInformation(const HdSceneIndexBaseRefPtr& viewportSceneIndex)
+{
+    //Block for the lifetime of the lock
+    {
+        std::lock_guard<std::mutex> lock(_viewportInformationAndSceneIndicesPerViewportDataSet_mutex);
+    
+        auto findResult = std::find_if(_viewportInformationAndSceneIndicesPerViewportDataSet.begin(), _viewportInformationAndSceneIndicesPerViewportDataSet.end(),
+                    [&viewportSceneIndex](const ViewportInformationAndSceneIndicesPerViewportData* other) { return other->GetViewportInformation()._viewportSceneIndex == viewportSceneIndex;});
+        if (findResult != _viewportInformationAndSceneIndicesPerViewportDataSet.end()){
+
+            RenderIndexProxy* renderIndexProxy = & (*findResult)->GetRenderIndexProxy();//Get the pointer on the renderIndexProxy
+
+            //Destroy the custom filtering scene indices chain
+            if (renderIndexProxy){
+                //Following code is equivalent to calling FilteringSceneIndicesChainManager::GetManager().DestroyFilteringSceneIndicesChain(*renderIndexProxy);
+                //But we cannot do so because of the lock above.
+                auto renderIndex = renderIndexProxy->GetRenderIndex();
+                if (renderIndex && (*findResult)->GetLastFilteringSceneIndexOfTheChain()){
+                    renderIndex->RemoveSceneIndex((*findResult)->GetLastFilteringSceneIndexOfTheChain());//Remove the whole chain from the render index
+                }
+            }
+            
+            delete (*findResult); //delete the ViewportInformationAndSceneIndicesPerViewportData*
```

Comment:
I started to store them by value as you suggested but there is still the ViewportInformation pointer to manage (new/delete) which is inside this class so we would still have to manage some memory. 
So is this something you still want ?

Reply:
Also since this is data per viewport and is not shared by any other class I think it should be manageable.

Reply:
I haven't managed to make it work with with unique_ptrs (there is a copy I can't find somewhere, but the error message only indicates lines in the standard library headers), but it does work with shared_ptrs. I'm not sure if the overhead would be a problem or not, but either ways, I think as long as we stick to the one `new` and the one `delete` we currently have, this should be fine as it's isolated in the implementation.

Reply:
(Grouping this conversation with [this comment](https://github.com/Autodesk/maya-hydra/pull/4#discussion_r1391763392))
So after taking a closer look, the situation with `ViewportInformation` objects is quite similar, as like you said, those are also manually managed, with the exception that objects of that class are expected to be manually created outside of the Flow Viewport Toolkit, and their ownership then transferred to the Flow Viewport Toolkit. I feel like this should also be solvable with unique_ptrs.

That being said, if we can't get smart pointers to work, then I think we should at least explicitly document that the `AddViewportInformation` method will assume ownership of the `ViewportInformation` object it receives, so that users know not to have the object they pass in be deleted.

Reply:
AddViewportInformation is not called by users, but by maya hydra only. But you're right, I changed the method signature and added more comments about the ownership.

Reply:
I am using a std::set with copies in it instead of arrays, the class to copy is not heavy and this set will have the size of the number of viewports in maya.